### PR TITLE
Add a bootstrap-haskell script, fixes #36

### LINKED ISSUE #36: Add get-ghcup installation script
The current process to install ghcup, ghc, and cabal requires several steps that could be automated. Other installers (such as [rustup](https://rustup.rs/)) have a very simple process of: `curl https://sh.rustup.rs -sSf | sh` which installs the `rustup` tool, `cargo` (package manager), and `rustc` (the compiler) all in one go. It would be really nice if we could do the same thing with ghcup :)

# Overview

I imagine the end-to-end user journey for someone that's never used Haskell before should be as simple as:

```bash
$ curl https://get.ghcup.haskell.org -sSf | sh
$ source $HOME/.ghcup/env
$ echo 'source $HOME/.ghcup/env' >> $HOME/.bashrc
$ mkdir myproject && cd myproject
$ cabal init -n --is-executable
$ cabal v2-run
```

# Details

To break things down a bit:

`curl https://get.ghcup.haskell.org -sSf | sh`

* Downloads and installs the most recent version of ghcup, ghc, and cabal
    * Separating out the ghc install and cabal install into follow-up work as described in the README seem like unnecessary steps to me, the first thing folks are going to do when they download it is run these commands. 
* If ghcup already exists it should detect this and offer to run `ghcup upgrade`
* Provide instructions for how to add the necessary bin directories to your path: `echo 'source $HOME/.ghcup/env' >> $HOME/.bashrc`

`source $HOME/.ghcup/env`

* The installation script above should create this env file containing: `export PATH="$HOME/.cabal/bin:$HOME/.ghcup/bin:$PATH"`

`echo 'source $HOME/.ghcup/env' >> $HOME/.bashrc`

* Make sure $PATH is set properly for future sessions
* This (and the line above) should be copy and paste from the output of the install script

# Implementation

I have a shell script that a wrote that accomplishes most of this now, it needs some polish but if you think this is a good idea I can send a PR and we can discuss more there.

### PR DESCRIPTION
**Pull request by hasufell**

### CODE REVIEW

**Comment by m-renaud** [comment]
I added a PR for the script I had previously written (#41) for comparison.

**Comment by hasufell** [comment]
@m-renaud please try the latest version

**Comment by hasufell** [comment]
travis is forever broken here and I cannot figure out how to fix it

**Comment by m-renaud** [comment]
@hasufell tried it out just now, works great other than the "wait for input not working".

I downloaded the script to a local file and ran `cat bootstrap-haskell | sh` and it printed the prompt to press enter to continue, but continued along with actually waiting for input.

If I change the script to be executable (`chmod +x bootstrap-haskell`) and then `./bootstrap-haskell` it works as expected. It looks like`read` has different behaviour if its piped into `sh` instead of being run as a script as part of a pipeline (see this [SO answer](https://stackoverflow.com/questions/16854041/bash-read-is-being-skipped-when-run-from-curl-pipe)).

So, it looks like our options are either:
1) Use the workaround provided there (3 commands, messier, -1 from me)
2) Don't prompt, just install.

My vote would be for (2), if they're on the ghcup site they already know what it does. At the bottom we could give instructions for how to uninstall ghcup (namely, `rm -rf ~/.ghcup`), or we could also provide a `ghcup uninstall` command which does the right thing. I think for an initial version just providing instructions would be sufficient.

Also, I noticed that the script runs `_edo ghcup set` which just prints the usage of the command at the bottom:

```
Done installing, run "ghci-8.4.4" or set up your current GHC via: ghcup set 8.4.4
ghcup set
ghcup-set
Set the currently active GHC to the specified version

USAGE:
    ghcup set [FLAGS] <VERSION>

FLAGS:
    -h, --help       Prints help information

ARGS:
    <VERSION>        E.g. "8.4.3" or "8.6.1"

DISCUSSION:
    Sets the the current GHC version by creating non-versioned
    symlinks for all ghc binaries of the specified version in
    "~/.ghcup/bin/<binary>".
```

Was that intentional?

File: .travis.yml [hunk 1]
```diff
@@ -15,16 +15,18 @@ addons:
     - gcc
     - git
     - libgmp-dev
-    - libllvm5.0
+    # - libllvm5.0
```
Comment (m-renaud) [inline .travis.yml]: Any reason why these and below are commented out?

**Comment by m-renaud** [comment]
I noticed you commented out some of the llvm deps in the travis script, what's the rationale behind that? It looks like it may be the cause of the issue.

**Comment by m-renaud** [comment]
Also, I just realized that by running just `ghcup set` returns a status code of `1` meaning that `cabal-install` is never installed.

File: .travis.yml (hunk 1 above)
Comment (hasufell) [reply .travis.yml]: > Any reason why these and below are commented out?

Because we don't use llvm code generator.

**Comment by hasufell** [comment]
> Also, I noticed that the script runs `_edo ghcup set` which just prints the usage of the command at the bottom:

That's because you are upgrading to an older version while the script is running. See what the travis.sh does.

> I noticed you commented out some of the llvm deps in the travis config, what's the rationale behind that? It looks like it may be the cause of the issue.

Why do you think that? It's gold linker f**kup afais and GHC just picks up whatever it wants.





> Also, I just realized that by running just `ghcup set` returns a status code of `1` meaning that `cabal-install` is never installed.

See above, you're upgrading to an older version while running the script. Set `BOOTSTRAP_HASKELL_NO_UPGRADE=1`.

**Comment by hasufell** [comment]
> If I change the script to be executable (`chmod +x bootstrap-haskell`) and then `./bootstrap-haskell` it works as expected. It looks like`read` has different behaviour if its piped into `sh` instead of being run as a script as part of a pipeline (see this [SO answer](https://stackoverflow.com/questions/16854041/bash-read-is-being-skipped-when-run-from-curl-pipe)).

Will look into it.

File: bootstrap-haskell [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	cat <<-EOF > "${HOME}"/.ghcup/env || _die "Failed to create env file"
+		export PATH="\$HOME/.cabal/bin:\$HOME/.ghcup/bin:\$PATH"
+		EOF
+	# shellcheck disable=SC1090
+	_edo . "${HOME}"/.ghcup/env
+fi
+
+_edo ghcup install
+
+_edo ghcup set
```
Comment (m-renaud) [inline bootstrap-haskell]: Replacing this by the following results in the script finishing successfully:

```
recommendedGhcVer=`ghcup list -t ghc | grep recommended | cut -f2`
_edo ghcup set $recommendedGhcVer
```

**Comment by hasufell** [comment]
> The bootstrap-haskell script has the line `_edo ghcup set` which is incorrect usage, causing the script to exit early and never run `_edo ghcup install-cabal` or any of the other commands below it.
> 
> Shouldn't it be `_edo ghcup set 8.4.4`?

No, the line is correct, I already explained it: You're upgrading to an older version while running the script. Set BOOTSTRAP_HASKELL_NO_UPGRADE=1. (see the commits of this PR)

**Comment by m-renaud** [comment]
> No, the line is correct...

Ah, yes you're right. I see where my confusion came from: I hadn't realized that there was a required change to `ghcup` in this PR to make `ghcup set` auto-discover the version to use if no arguments are passed. I had just pulled down a copy of the bootstrap script, not the updated `ghcup` so the `ghcup` that was being downloaded and used obviously didn't have this change so was erroring out. Sorry about the noise.

File: .travis.yml (hunk 1 above)
Comment (m-renaud) [reply .travis.yml]: Maybe I should have commented on the lines below, but `llvm-5.0` and `llvm-5.0-dev` are both removed, and afaik at least one of those provides a linker in the environment. You mentioned in another comment that Travis is broken, I wouldn't be entirely surprised if this was why.

Comment (hasufell) [reply .travis.yml]: I doubt it. llvm is only used for llvm code generation. The linker is provided by binutils.

**Comment by hasufell** [comment]
So, in order to merge this I would basically have to exclude this from the travis test...

**Comment by m-renaud** [comment]
I think it would be reasonable to exclude this from the Travis test initially and figure how to add it in as a follow-up.

**Comment by hasufell** [comment]
The only issue remaining was the `read` thing right?

**Comment by m-renaud** [comment]
Yup, I think a reasonable solution for the first version of the script is to just do it without asking.

**Comment by hasufell** [comment]
> If I change the script to be executable (`chmod +x bootstrap-haskell`) and then `./bootstrap-haskell` it works as expected. It looks like`read` has different behaviour if its piped into `sh` instead of being run as a script as part of a pipeline (see this [SO answer](https://stackoverflow.com/questions/16854041/bash-read-is-being-skipped-when-run-from-curl-pipe)).
> 
> So, it looks like our options are either:
> 
>     1. Use the workaround provided there (3 commands, messier, -1 from me)
> 
>     2. Don't prompt, just install.

I explicitly read from `/dev/tty` now. This seems to work well. I'm not 100% sure how portable that is, but it seems to be part of posix: http://pubs.opengroup.org/onlinepubs/009696699/basedefs/xbd_chap10.html

**Pull request merged**